gslib/wildcard_iterator.py

```diff
@@ -649,8 +651,12 @@ def __iter__(self, bucket_listing_fields=None):
     for filepath in filepaths:
       expanded_url = StorageUrlFromString(filepath)
       try:
-        if self.exclude_pattern is not None:
-          if self.exclude_pattern.match(str(expanded_url)):
+        if self.exclude_tuple is not None:
```

Comment:
Should the code from here be removed, now that we are filtering out things at iterator level? https://github.com/GoogleCloudPlatform/gsutil/blob/2fd97591681a51ca0541d04b865e7d67a54efad4/gslib/commands/rsync.py#L764

Reply:
Yes it should...

Reply:
It turns out we need both. It turns out `wildcard_iterator` is excluding directories while `rsync` is excluding files. I've added a comment to this effect to make it more clear.

Reply:
Perhaps the bug mentioned in the comment below was the reason it was needed in rsync. After the fix, I think it should be fine to avoid doing the filtering in `rsync.py`? However, given that we are only doing this in wildcard iterator for file urls, we might still want to use the exclude patterns for cloud urls. 

I think at some point we should move the filtering for cloud urls inside wildcard iterator as well, but that might not be worth doing in gsutil, maybe for gcloud storage. 

For now, I think we can do this in `rsync.py`?
```
if cls.exclude_pattern and not url.IsFileUrl():
  ... 
   ...
```